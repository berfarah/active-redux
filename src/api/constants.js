// API
export const API_CONFIGURE = 'AR_API_CONFIGURE';

// REMOTE
export const REMOTE_CLEAR = 'AR_REMOTE_CLEAR';
export const REMOTE_HYDRATE = 'REMOTE_HYDRATE';

export const REMOTE_WILL_READ = 'AR_REMOTE_WILL_READ';
export const REMOTE_READ_DONE = 'AR_REMOTE_READ_DONE';
export const REMOTE_READ_FAILED = 'AR_REMOTE_READ_FAILED';

export const REMOTE_WILL_CREATE = 'AR_REMOTE_WILL_CREATE';
export const REMOTE_CREATE_DONE = 'AR_REMOTE_CREATE_DONE';
export const REMOTE_CREATE_FAILED = 'AR_REMOTE_CREATE_FAILED';

export const REMOTE_WILL_UPDATE = 'AR_REMOTE_WILL_UPDATE';
export const REMOTE_UPDATE_DONE = 'AR_REMOTE_UPDATE_DONE';
export const REMOTE_UPDATE_FAILED = 'AR_REMOTE_UPDATE_FAILED';

export const REMOTE_WILL_DELETE = 'AR_REMOTE_WILL_DELETE';
export const REMOTE_DELETE_DONE = 'AR_REMOTE_DELETE_DONE';
export const REMOTE_DELETE_FAILED = 'AR_REMOTE_DELETE_FAILED';
