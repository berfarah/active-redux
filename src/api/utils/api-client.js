import axios from 'axios';
import createError from 'axios/lib/core/createError';

const jsonContentTypes = [
  'application/json',
  'application/vnd.api+json',
];

const validateContentType = (response) => {
  if (jsonContentTypes.includes(response.headers['content-type'])) return;

  throw createError(
    'Invalid Content-Type in response',
    response.config,
    null,
    response
  );
};

const mergeDefaults = (url, options) => ({
  ...options,
  url,
  headers: {
    Accept: 'application/vnd.api+json',
    'Content-Type': 'application/vnd.api+json',
    ...options.headers,
  },
});

export default (url, options = {}) => (
  axios(mergeDefaults(url, options)).then((response) => {
    if (response.status === 204) return response;
    validateContentType(response);
    return response.data;
  })
);
