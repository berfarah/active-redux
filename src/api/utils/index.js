import imm from 'object-path-immutable';

export apiClient from './api-client';

export const incrementProperty = (state, key) => (
  imm.set(state, key, state[key] + 1)
);

export const decrementProperty = (state, key) => (
  imm.set(state, key, state[key] - 1)
);

export const resourcesArray = (data) => (
  Array.isArray(data) ? data : [data]
);

export const mergeResources = (state, { data, included = [] }) => {
  const newState = imm(state);

  resourcesArray(data).concat(included).forEach((dataObj) => {
    // if we don't do this and the ID is a number, it'll create an array
    newState.set(dataObj.type, state[dataObj.type] || {});
    newState.set(`${dataObj.type}.${dataObj.id}`, dataObj);
  });

  return newState.value();
};

export const clearResources = (state, data) => {
  switch (typeof data) {
    case 'object': {
      if (Array.isArray(data)) return data.reduce(clearResources, state);
      const { type, id } = data;
      return imm.del(state, ['resources', type, id]);
    }
    case 'string': return imm.set(state, ['resources', data], {});
    case 'undefined': return imm.set(state, 'resources', {});
    default: return state;
  }
};

export const createAction = (type) => (payload) => ({ type, payload });

export const createReducer = (mappings, initialState) => (state = initialState, action) => (
  mappings[action.type] ? mappings[action.type](state, action) : state
);

export const hasProperty = (object, ...keys) => (
  keys.reduce((acc, key) => {
    if (acc === undefined) return acc;
    return acc[key];
  }, object)
);

export const markPendingResources = (state, { data }) => (
  resourcesArray(data).reduce((acc, resource) => {
    if (!hasProperty(state, 'resources', resource.type, resource.id)) return acc;
    return acc.set(['resources', resource.type, resource.id, 'isPending'], true);
  }, imm(state)).value()
);
