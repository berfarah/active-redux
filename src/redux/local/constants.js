export const LOCAL_CLEAR = 'AR_LOCAL_CLEAR';
export const LOCAL_HYDRATE = 'AR_LOCAL_HYDRATE';
export const LOCAL_READ = 'AR_LOCAL_READ';
export const LOCAL_CREATE = 'AR_LOCAL_CREATE';
export const LOCAL_UPDATE = 'AR_LOCAL_UPDATE';
export const LOCAL_DELETE = 'AR_LOCAL_DELETE';
