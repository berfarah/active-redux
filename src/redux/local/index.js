export * from './actions';
export localReducer from './reducer';
