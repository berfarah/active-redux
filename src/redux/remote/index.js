export * from './actions';
export remoteReducer from './reducer';
