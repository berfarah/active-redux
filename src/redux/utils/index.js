import imm from 'object-path-immutable';

export apiClient from './api-client';

export const incrementProperty = (state, key) => (
  imm.set(state, key, state[key] + 1)
);

export const decrementProperty = (state, key) => (
  imm.set(state, key, state[key] - 1)
);

export const resourcesArray = (data) => (
  Array.isArray(data) ? data : [data]
);

export const mergeResources = (state, { data, included = [] }) => {
  const newState = imm(state);

  resourcesArray(data).concat(included).forEach((dataObj) => {
    // if we don't do this and the ID is a number, it'll create an array
    newState.set(dataObj.type, state[dataObj.type] || {});
    newState.set(`${dataObj.type}.${dataObj.id}`, dataObj);
  });

  return newState.value();
};

export const clearResources = (state, resources) => {
  switch (typeof resources) {
    case 'object': {
      if (Array.isArray(resources)) return resources.reduce(clearResources, state);
      const { type, id } = resources;
      return imm.del(state, ['resources', type, id]);
    }
    case 'string': return imm.set(state, ['resources', resources], {});
    case 'undefined': return imm.set(state, 'resources', {});
    default: return state;
  }
};

export const createAction = (type) => (payload) => ({ type, payload });

export const createReducer = (mappings, initialState) => (state = initialState, action) => (
  mappings[action.type] ? mappings[action.type](state, action) : state
);
