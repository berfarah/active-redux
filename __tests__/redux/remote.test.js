import * as subject from 'active-redux/redux/remote';
import * as ActionTypes from 'active-redux/redux/remote/constants';
import { createAction } from 'active-redux/redux/utils';

describe('remoteReducer', () => {
  const person = {
    type: 'people',
    id: 5,
    attributes: {
      hair: 'black',
      age: 25,
    }
  };

  const fullStore = {
    apiConfig: {},
    isCreating: 0,
    isDeleting: 0,
    isReading: 0,
    isUpdating: 0,
    resources: {
      [person.type]: {
        [person.id]: { attributes: { name: 'Johnny' } },
        2: { attributes: { name: 'Yangster' } }
      }
    }
  };

  describe('initialState', () => {
    it('has an empty resources object', () => {
      expect(subject.remoteReducer(undefined, {})).toMatchSnapshot();
    });
  });

  describe('actions', () => {
    const reduce = subject.remoteReducer;

    describe('apiConfigure()', () => {
      it('updates the apiConfig key', () => {
        const state = reduce(undefined, subject.apiConfigure({ baseUrl: 'http://coverhound.com/api' }));
        expect(state).toMatchSnapshot();
      });
    });


    describe('remoteHydrate()', () => {
      it('populates the state with resources', () => {
        const state = reduce(undefined, subject.remoteHydrate({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('remoteCreate()', () => {
      let state = fullStore;
      const remoteWillCreate = createAction(ActionTypes.REMOTE_WILL_CREATE);
      const remoteCreateDone = createAction(ActionTypes.REMOTE_CREATE_DONE);

      it('increments isCreating', () => {
        state = reduce(undefined, remoteWillCreate({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        state = reduce(state, remoteCreateDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('remoteRead()', () => {
      let state = fullStore;
      const remoteWillRead = createAction(ActionTypes.REMOTE_WILL_READ);
      const remoteReadDone = createAction(ActionTypes.REMOTE_READ_DONE);

      it('increments isReading', () => {
        state = reduce(undefined, remoteWillRead({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('populates the state with resources', () => {
        state = reduce(state, remoteReadDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('remoteClear()', () => {
      it('clears the resource', () => {
        const state = reduce(fullStore, subject.remoteClear(person.type));
        expect(state).toMatchSnapshot();
      });
    });

    describe('remoteUpdate()', () => {
      let state = fullStore;
      const remoteWillUpdate = createAction(ActionTypes.REMOTE_WILL_UPDATE);
      const remoteUpdateDone = createAction(ActionTypes.REMOTE_UPDATE_DONE);

      it('increments isUpdating and marks the resource as pending', () => {
        state = reduce(state, remoteWillUpdate({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('updates the state with resources', () => {
        state = reduce(state, remoteUpdateDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });

    describe('remoteDelete()', () => {
      let state = fullStore;
      const remoteWillDelete = createAction(ActionTypes.REMOTE_WILL_DELETE);
      const remoteDeleteDone = createAction(ActionTypes.REMOTE_DELETE_DONE);

      it('increments isDeleting and marks the resource as pending', () => {
        state = reduce(state, remoteWillDelete({ data: person }));
        expect(state).toMatchSnapshot();
      });

      it('removes the resources from the store', () => {
        state = reduce(state, remoteDeleteDone({ data: person }));
        expect(state).toMatchSnapshot();
      });
    });
  });
});
