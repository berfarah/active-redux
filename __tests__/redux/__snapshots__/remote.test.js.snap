// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remoteReducer actions apiConfigure() updates the apiConfig key 1`] = `
Object {
  "apiConfig": Object {
    "baseUrl": "http://coverhound.com/api",
  },
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`remoteReducer actions remoteClear() clears the resource 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {},
  },
}
`;

exports[`remoteReducer actions remoteCreate() increments isCreating 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 1,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`remoteReducer actions remoteCreate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`remoteReducer actions remoteDelete() increments isDeleting and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 1,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "name": "Johnny",
        },
        "isPending": true,
      },
    },
  },
}
`;

exports[`remoteReducer actions remoteDelete() removes the resources from the store 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
    },
  },
}
`;

exports[`remoteReducer actions remoteHydrate() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`remoteReducer actions remoteRead() increments isReading 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 1,
  "isUpdating": 0,
  "resources": Object {},
}
`;

exports[`remoteReducer actions remoteRead() populates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`remoteReducer actions remoteUpdate() increments isUpdating and marks the resource as pending 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 1,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "name": "Johnny",
        },
        "isPending": true,
      },
    },
  },
}
`;

exports[`remoteReducer actions remoteUpdate() updates the state with resources 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {
    "people": Object {
      "2": Object {
        "attributes": Object {
          "name": "Yangster",
        },
      },
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`remoteReducer initialState has an empty resources object 1`] = `
Object {
  "apiConfig": Object {},
  "isCreating": 0,
  "isDeleting": 0,
  "isReading": 0,
  "isUpdating": 0,
  "resources": Object {},
}
`;
